task 8

Shreksatronik opened the pull request.

Shreksatronik committed:
task 8

trushev commented on task 8/Counter.java:
```diff
@@ -0,0 +1,19 @@
+public class Counter {
+    private static final int N = 10_000_000;
```
Эта задача сложнее, чем вы подумали
Смысл в том, что программа должна работать до тех пор, пока пользователь её не прервёт
`N` вообще не должно быть в коде

trushev commented on task 8/Main.java:
```diff
@@ -0,0 +1,36 @@
+            @Override
+            public void run() {
+                System.out.println("Shutdown hook ran!");
+                new Barriers(counters).run();
```
В этот момент потоки `PiThread` должны еще работать по условию задачи, без синхронизации(или volatile) с ними вы прочитаете из переменных `sum`, скорее всего, нули

Commit:
added changes

Commit:
added changes

Как запускать: скомпилировать исходники javac  Main.java Counter.java
java Main и затем ввести количество тредов

trushev commented on task 8/src/main/java/Main.java:
```diff
@@ -0,0 +1,46 @@
+                    counters.get(i).stop();
+                }
+                try {
+                    Thread.sleep(2000);
```
Я надеялся, что вы к моменту решения 8 задачи уже осознали, что Thread.sleep не работает для синхронизации потоков

trushev commented on task 8/src/main/java/Main.java:
```diff
@@ -0,0 +1,46 @@
+            public void run() {
+                System.out.println("Is working");
+                for (int i = 0; i < Count; i++) {
+                    counters.get(i).stop();
```
Уже писал про это https://github.com/Shreksatronik/JavaDB/pull/2#discussion_r989823462

trushev commented on task 8/src/main/java/Counter.java:
```diff
@@ -0,0 +1,47 @@
+                }
+            }
+            try {
+                Thread.sleep(1);
```
Зачем?

trushev commented on task 8/src/main/java/Counter.java:
```diff
@@ -0,0 +1,47 @@
+            i++;
+            if (i % 100 == 0){
+                try{
+                    cyclicBarrier.await();
```
Способ работает, но в этой задаче не оправдан
Подумайте, как не синхронизовать `Counter` между собой во время вычислений. Синхронизация понадобится только в момент работы `shutdownHook`

trushev commented on task 8/src/main/java/Counter.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Override
+    public void run() {
+        int iterator = threadNum;
```
`int` маловат для этой цели, быстро наступит переполение

Commit:
added changes

Shreksatronik committed:
Delete task 8 directory

Shreksatronik committed:
added changes

Shreksatronik committed:
Add files via upload

Shreksatronik committed:
Add files via upload

Shreksatronik committed:
Add files via upload
